### PR TITLE
merge stable to release 1.1

### PR DESCRIPTION
**Pull request by ajtmccarty**
had to make a separate branch to address how some diff query changes on `stable` interact with other diff changes on `release-1.1`

had to make more changes than expected to the diff calculation logic to interact correctly with a bug that I fixed in `stable`. fortunately, we have a lot of tests that cover the diff pretty well

### CODE REVIEW

File: backend/infrahub/core/query/diff.py [hunk 1]
```diff
@@ -668,7 +668,7 @@ async def query_init(self, db: InfrahubDatabase, **kwargs: Any) -> None:
             AND (from_time <= diff_rel.from < $to_time)
             AND (diff_rel.to IS NULL OR (from_time <= diff_rel.to < $to_time))
             AND r_root.from <= diff_rel.from
-            AND (r_root.to IS NULL OR r_root.to >= diff_rel.from)
+            AND (r_root.to IS NULL OR diff_rel.branch <> r_root.branch OR r_root.to >= diff_rel.from)
```
Comment (ajtmccarty) [inline backend/infrahub/core/query/diff.py:671]: this little change is required to account for this case
1. make branch `branch`
2. delete `NodeA` on `main`
3. update a relationship to `NodeA` on `branch`, where `NodeA` is not deleted

File: backend/infrahub/core/diff/query_parser.py [hunk 2]
```diff
@@ -365,12 +365,11 @@ def to_diff_relationship(self, from_time: Timestamp, include_unchanged: bool) ->
...
-            self.cardinality is RelationshipCardinality.ONE
-            and len(single_relationships) == 1
-            and single_relationships[0].action in (DiffAction.ADDED, DiffAction.REMOVED)
-        ):
-            action = single_relationships[0].action
+        # bubble up action, excluding UNCHANGED
+        if self.cardinality is RelationshipCardinality.ONE:
+            actions = {sr.action for sr in single_relationships if sr.action is not DiffAction.UNCHANGED}
+            if len(actions) == 1:
+                action = actions.pop()
```
Comment (ajtmccarty) [inline backend/infrahub/core/diff/query_parser.py:374]: ignore `UNCHANGED` individual relationships when determining the overall action for this relationship group (kind of equivalent to RelationshipManager)

File: backend/tests/integration/diff/test_diff_merge.py [hunk 3]
```diff
@@ -146,11 +146,11 @@ async def test_node_delete_conflict(
...
         diff_branch = await create_branch(db=db, branch_name="branch2")
         if delete_on_branch:
-            delete_branch = default_branch
-            update_branch = diff_branch
-        else:
             delete_branch = diff_branch
             update_branch = default_branch
+        else:
+            delete_branch = default_branch
+            update_branch = diff_branch
```
Comment (ajtmccarty) [inline backend/tests/integration/diff/test_diff_merge.py:153]: the naming and logic was inverted here, which confused me while investigating this test

File: backend/infrahub/core/diff/query_parser.py [hunk 4]
```diff
@@ -583,14 +604,25 @@ def _update_attribute_property(
...
+            )
+            branch_name = database_path.deepest_branch
+            from_time = self.from_time
+            if branch_name == self.base_branch_name and node_field_specifier in self.get_new_node_field_specifiers():
+                from_time = self.diff_branched_from_time
             diff_relationship = DiffRelationshipIntermediate(
                 name=relationship_schema.name,
                 cardinality=relationship_schema.cardinality,
                 identifier=relationship_schema.get_identifier(),
+                from_time=from_time,
```
Comment (ajtmccarty) [inline backend/infrahub/core/diff/query_parser.py:625]: we use the `from_time` here and above in `_get_diff_attribute` when determining the action (ADDED/UPDATED/REMOVED) for a given database path. when we are getting nodes on the base branch (always `main` for now), we retrieve both completely new nodes that are not yet part of this diff and updated nodes that are already part of this diff. we need to use different `from_time`s when determining their actions and this is how we accomplish that

File: backend/infrahub/core/diff/query_parser.py [hunk 5]
```diff
@@ -487,6 +490,16 @@ def get_node_field_specifiers_for_branch(self, branch_name: str) -> set[NodeFiel
...
 
+    def get_new_node_field_specifiers(self) -> set[NodeFieldSpecifier]:
+        branch_node_specifiers = self.get_diff_node_field_specifiers()
+        new_node_field_specifiers = branch_node_specifiers - self._previous_node_field_specifiers
+        return new_node_field_specifiers
+
+    def get_current_node_field_specifiers(self) -> set[NodeFieldSpecifier]:
+        new_node_field_specifiers = self.get_new_node_field_specifiers()
+        current_node_field_specifiers = self._previous_node_field_specifiers - new_node_field_specifiers
+        return current_node_field_specifiers
```
Comment (ajtmccarty) [inline backend/infrahub/core/diff/query_parser.py:501]: I'd like to add some caching for these b/c we should really only need to calculate these things one time, but that can come in a later PR

**Review by dgarros — APPROVED** [review]

**Pull request merged**